packages/lib/src/actionMiddlewares/draftMiddleware.ts

```diff
@@ -0,0 +1,102 @@
+import { getParentToChildPath, resolvePath } from "../parent"
+import { applyPatches, onPatches, OnPatchesDisposer, patchRecorder } from "../patch"
+import { isRootStore, registerRootStore, unregisterRootStore } from "../rootStore"
+import { clone } from "../snapshot"
+import { assertTweakedObject } from "../tweaker/core"
+
+/**
+ * Callback function for `DraftManager.withDraft`.
+ */
+export type WithDraftCallback<T extends object> = (draftNode: T) => boolean | undefined | void
+
+/**
+ * Manager class returned by `draftMiddleware` that allows you to make changes to a draft copy of
+ * the original subtree and apply them to the original subtree or reject them.
+ */
+export class DraftManager {
+  /**
+   * The draft copy of the original subtree.
+   */
+  private readonly subtreeRootDraft: object
+
+  /**
+   * The internal `onPatches` disposer.
+   */
+  private readonly disposer: OnPatchesDisposer
+
+  /**
+   * Creates an instance of `DraftManager`.
+   * Do not use directly, use `draftMiddleware` instead.
+   *
+   * @param subtreeRoot Subtree root target object.
+   */
+  public constructor(private readonly subtreeRoot: object) {
+    assertTweakedObject(subtreeRoot, "subtreeRoot")
+    this.subtreeRootDraft = clone(subtreeRoot)
+    if (isRootStore(subtreeRoot)) {
+      registerRootStore(this.subtreeRootDraft)
+    }
+    this.disposer = onPatches(subtreeRoot, patches => applyPatches(this.subtreeRootDraft, patches))
+  }
+
+  /**
+   * Executes `fn` with a draft of `node`. The changes made to the draft in `fn` can be accepted,
+   * i.e. applied to the original subtree, or rejected.
+   *
+   * @typeparam T Object type.
+   * @param node Object for which to obtain a draft copy.
+   * @param fn Function that is caled with a draft copy of `node`. Any changes made to the draft are
+   * applied to the original subtree when `fn` returns `true`. When `fn` returns `false`,
+   * `undefined` or nothing, the changes made to the draft are rejected.
+   */
+  public withDraft<T extends object>(node: T, fn: WithDraftCallback<T>): void {
+    assertTweakedObject(node, "node")
+
+    const path = getParentToChildPath(this.subtreeRoot, node)
+    if (!path) {
+      return
+    }
+
+    const draftNode = resolvePath<T>(this.subtreeRootDraft, path).value
+    if (!draftNode) {
+      return
```

Comment:
maybe `throw failure("...")`

Reply:
After looking at this again, I think an error check is not needed at all because the path must be resolvable. Do you agree?

Reply:
Wouldn't it fail if for some reason the draft is out of sync with the original?

Reply:
You're right. Ideally, this should never happen, but I've added the check in e79e4656b26297cea09a316b6a8d74e706e64a70 to be on the safe side.